microsoft/testsuites/dpdk/common.py

```diff
@@ -17,6 +22,261 @@
 AZ_ROUTE_ALL_TRAFFIC = "0.0.0.0/0"
 
 
+# Attempt to clean up the DPDK package dependency mess
+# Make a Installer class that implements the common steps
+# for installing DPDK/rdma-core, either from source or the package manager.
+# This generic class will get implemented in DpdkTestpmd and RdmaCore.
+# This should help us cover the various installation cases in a nice way,
+# and allow us to only re-implement the bits we need for each project.
+class OsPackageDependencies:
+    # A class to reduce the isinstance() trees that are
+    # sprinkled everywhere.
+    # Caller provides a function to match an OS and
+    # the packages to install on that OS.
+    def __init__(
+        self,
+        os_matcher: Callable[[Posix], bool],
+        packages: Sequence[Union[str, Tool, Type[Tool]]],
+    ) -> None:
+        self.check_node_os = os_matcher
+        self.packages = packages
+
+
+class DependencyInstaller:
+    # provide a list of OsPackageDependencies for a project
+    def __init__(self, requirements: List[OsPackageDependencies]) -> None:
+        self.requirements = requirements
+
+    # evaluate the list of package dependencies,
+    def install_required_packages(
+        self, node: Node, extra_args: Union[List[str], None]
+    ) -> None:
+        os = node.os
+        if not isinstance(os, Posix):
+            return
+        # find the first match for an OS, install the packages.
+        # Most specific should go first, most generic should go last.
+        for os_package_check in self.requirements:
+            if os_package_check.check_node_os(os):
+                os.install_packages(os_package_check.packages, extra_args=extra_args)
+                return
+        raise UnsupportedDistroException(
+            os,
+            message=("Installer did not define dependencies for this os."),
+        )
+
+
+class Downloader:
+    def __init__(self, node: Node) -> None:
+        self._node = node
+
+    def download(self) -> PurePath:
+        raise NotImplementedError("Downloader not implemented.")
+
+
+class GitDownload(Downloader):
+    git_repo: str = ""
+    git_ref: str = ""
+
+    def __init__(
+        self,
+        node: Node,
+        git_repo: str,
+        git_ref: str,
+    ) -> None:
+        super().__init__(node)
+        self.git_repo = git_repo
+        self.git_ref = git_ref
+
+    # checkout the git repository into the working path
+    def download(self) -> PurePath:
+        # NOTE: fail on exists is set to True.
+        # The expectation is that the parent Installer class should
+        # remove any lingering installations
+        self._source_path = self._node.tools[Git].clone(
+            self.git_repo,
+            cwd=self._node.get_working_path(),
+            ref=self.git_ref,
+            fail_on_exists=True,
+        )
+        self._branch_handler()
```

Comment:
It doesn't need to define another method. When it needs extra operation, just overwrites the download method in the subclass like below. It gives more flexibility. 
```python
def download(self) -> PurePath:
    # other operations before download.
    super().download()
    # other operations after download.
```

Reply:
Ok